cmd/sops/common/common.go

```diff
@@ -136,3 +140,178 @@ func DefaultStoreForPath(path string) Store {
+}
+
+type GenericDecryptOpts struct {
+	Cipher      sops.Cipher
+	InputStore  sops.Store
+	InputPath   string
+	IgnoreMAC   bool
+	KeyServices []keyservice.KeyServiceClient
+}
+
+func LoadEncryptedFileWithBugFixes(opts GenericDecryptOpts) (*sops.Tree, error) {
+	tree, err := LoadEncryptedFile(opts.InputStore, opts.InputPath)
+	if err != nil {
+		return nil, err
+	}
+
+	encCtxBug, err := DetectKMSEncryptionContextBug(tree)
+	if err != nil {
+		return nil, err
+	}
+	if encCtxBug {
+		tree, err = FixAWSKMSEncryptionContextBug(opts, tree)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return tree, nil
+}
+
+func FixAWSKMSEncryptionContextBug(opts GenericDecryptOpts, tree *sops.Tree) (*sops.Tree, error) {
+	// TODO - Finalize messaging here.
+	message := "Up until version 3.3.0 of sops there was a bug surrounding the " +
+		"use of encryption context with AWS KMS."
+	fmt.Println(wordwrap.WrapString(message, 75))
+
+	if !terminal.IsTerminal(int(os.Stdout.Fd())) {
+		return nil, NewExitError("Need to run this from a tty to resolve this issue.", codes.TTYRequired)
+	}
+
+	var response string
+	for response != "y" && response != "n" {
+		fmt.Println("Would you like sops to automatically fix this issue? (y/n): ")
+		_, err := fmt.Scanln(&response)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if response == "n" {
+		return nil, fmt.Errorf("Exiting. User responded no.")
+	}
+
+	dataKey := []byte{}
+	success := false
+	// If there is another key, then we should be able to just decrypt
+	// without having to try different variations of the encryption context.
+	dataKey, err := DecryptTree(DecryptTreeOpts{
+		Cipher:      opts.Cipher,
+		IgnoreMac:   opts.IgnoreMAC,
+		Tree:        tree,
+		KeyServices: opts.KeyServices,
+	})
+	if err != nil {
+		kgndx, kndx, originalKey := GetKMSKeyWithEncryptionCtx(tree)
+
+		keyToEdit := *originalKey
+
+		encCtxVals := []string{}
+		for _, v := range keyToEdit.EncryptionContext {
+			encCtxVals = append(encCtxVals, *v)
+		}
+
+		encCtxVariations := []map[string]*string{}
+		for _, ctxVal := range encCtxVals {
+			encCtxVariation := map[string]*string{}
+			for key := range keyToEdit.EncryptionContext {
+				val := ctxVal
+				encCtxVariation[key] = &val
+			}
+			encCtxVariations = append(encCtxVariations, encCtxVariation)
+		}
+
+		for _, encCtxVar := range encCtxVariations {
+			keyToEdit.EncryptionContext = encCtxVar
+			tree.Metadata.KeyGroups[kgndx][kndx] = &keyToEdit
+			dataKey, err = DecryptTree(DecryptTreeOpts{
+				Cipher:      opts.Cipher,
+				IgnoreMac:   opts.IgnoreMAC,
+				Tree:        tree,
+				KeyServices: opts.KeyServices,
+			})
+			if err == nil {
+				success = true
+				tree.Metadata.KeyGroups[kgndx][kndx] = originalKey
+				tree.Metadata.Version = version.Version
+				break
+			}
+		}
+	} else {
+		success = true
+	}
+
+	if !success {
+		return nil, NewExitError("Failed to decrypt, meaning there is likely another problem from the encryption context bug.", codes.ErrorDecryptingTree)
```

Comment:
It'd be nice to show the user what the error is here

Reply:
What is the recommended way to do this within `sops` cli? Just `fmt.Println` it?

Reply:
The simple solution would be to just wrap it, like:

```go
NewExitError(fmt.Sprintf("failed to decrypt, meaning there is likely another problem from the encryption context bug: %s", err), codes.ErrorDecryptingTree)
```

The more complicated solution would be implementing a [`UserError`](https://github.com/mozilla/sops/blob/master/usererrors.go#L15) for it. Or implementing some sort of `UserError` wrapping behavior to avoid having to make its own type for it.

Reply:
Oh gotcha, yeah I completely misunderstood what you meant. Fixing.